New API for XAML group transformers and AvaloniaRuntimeXamlLoader.LoadGroup

maxkatz6 opened the pull request:
## What does the pull request do?

Partial reimplementation of my [previous PR](https://github.com/AvaloniaUI/Avalonia/pull/9413) with bunch of new features (mostly for internal use).

## What is the current behavior?

There was no way to:
- For any feature requiring transforming of nodes with information of all XAML documents in the project, it was necessary to use Cecil IL code emit, after main XAML files were already emitted.
- Easily handle debug information for these transformations
- Unit test combination multiple XAML files without mocked asset loader hacks
- As a result of above, StyleInclude/ResourceInclude transformers aren't really usable in current master

## What is the updated/expected behavior with this PR?

It is possible now to:
- Process XAML documents together after they all were parsed and transformed, but before they were emitted.
- Debug information is mostly handled automatically, as there is no unnecessary IL rewrite after
- `AvaloniaRuntimeXamlLoader.LoadGroup` allows to compile and parse multiple files and test their combination.
- Known StyleInclude/ResourceInclude bugs are fixed

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes

Comparing to previous master builds (but not preview), there are couple of binary breaking changes.

Also, in the public API some changes were made for the AvaloniaRuntimeXamlLoader. Some properties were moved from RuntimeXamlLoaderConfiguration to the RuntimeXamlLoaderDocument, as they were specific to a single file, while same configuration should be applicable to a group of files.
New `Load` method has following signature:
```
static object Load(RuntimeXamlLoaderDocument document, RuntimeXamlLoaderConfiguration? configuration = null);
static IReadOnlyList<object> LoadGroup(IReadOnlyCollection<RuntimeXamlLoaderDocument> documents, RuntimeXamlLoaderConfiguration? configuration = null);
```


## Fixed issues

Fixes https://github.com/AvaloniaUI/Avalonia/issues/9127

maxkatz6 committed:
Implement multiple documents loading and transforming in AvaloniaRuntimeXamlLoader

maxkatz6 committed:
Support group transformations in the XAML Build Task

maxkatz6 committed:
Make AvaloniaXamlIncludeTransformer a group transformer and remove old code with manual IL emit

maxkatz6 committed:
Update tests with "LoadGroup" method instead of assetloader hacks

maxkatz6 committed:
Fix build of BidningDemo

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0026553-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

maxkatz6 committed:
Minor cleanup

maxkatz6 commented on src/Avalonia.Build.Tasks/XamlCompilerTaskExecutor.cs:
```diff
@@ -370,15 +403,11 @@ bool CompileGroup(IResourceGroup group)
                                         {
-                                            foundXamlLoader = true;
-                                            break;
+                                            throw new InvalidProgramException("Same XAML file was loaded twice." +
+                                                "Make sure there is no x:Class duplicates no files were added to the AvaloniaResource msbuild items group twice.");
```
@kekekeks as I understand, we don't need this hack anymore.

kekekeks replied:
IIRC it was required for compatibility with pre-`.axaml` themes that was achieved via including the same file as EmbeddedResource and AvaloniaResource.

maxkatz6 committed:
Use XamlX fork with trimming fix

maxkatz6 committed:
Add missed RuntimeIdentifier to the xaml external build task

maxkatz6 committed:
Use main XamlX again

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0026567-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

maxkatz6 commented on packages/Avalonia/AvaloniaBuildTasks.targets:
```diff
@@ -74,7 +74,7 @@
     <Exec 
       Condition="'$(_AvaloniaUseExternalMSBuild)' == 'true'"
-      Command="dotnet msbuild /nodereuse:false $(MSBuildProjectFile) /t:GenerateAvaloniaResources /p:_AvaloniaForceInternalMSBuild=true /p:Configuration=$(Configuration) /p:TargetFramework=$(TargetFramework) /p:BuildProjectReferences=false"/>
+      Command="dotnet msbuild /nodereuse:false $(MSBuildProjectFile) /t:GenerateAvaloniaResources /p:_AvaloniaForceInternalMSBuild=true /p:Configuration=$(Configuration) /p:TargetFramework=$(TargetFramework) /p:RuntimeIdentifier=$(RuntimeIdentifier) /p:BuildProjectReferences=false"/>
```
This was causing compile errors with if I change xaml project target framework to "net7.0".
Main msbuild task was running with a runtime identifier set (by default in net7.0), while external msbuild task wasn't (no idea why, haven't checked too deep). 
As the result, IntermediateOutputPath was different between these two runs, and resource files were missing.

jmacato committed:
Merge branch 'master' into xaml-group-transfomers

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0026596-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

kekekeks commented on src/Markup/Avalonia.Markup.Xaml.Loader/AvaloniaRuntimeXamlLoader.cs:
```diff
@@ -17,31 +18,15 @@ public static class AvaloniaRuntimeXamlLoader
-        /// <param name="xaml">The string containing the XAML.</param>
-        /// <param name="configuration">Xaml loader configuration.</param>
-        /// <returns>The loaded object.</returns>
-        public static object Load(string xaml, RuntimeXamlLoaderConfiguration configuration)
```
IIRC it was supposed to be a public API for users who want to load XAML at runtime.

kekekeks replied:
So the idea was to have configuration parameters in a separate class so we can add more of them without breaking compatibility.

maxkatz6 replied:
Yes, I added this configuration parameter couple of weeks ago. So no breaking changes here.
But this configuration shouldn't contain any file source related configs, like RootObject, otherwise it won't be reusable in the LoadGroup methods, where there could be multiple root objects.

kekekeks commented on src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/GroupTransformers/XamlIncludeGroupTransformer.cs:
```diff
@@ -0,0 +1,133 @@
+            return node;
+        }
+
+        if (originalAssetPath.StartsWith("avares://"))
```
I think it would be more readable as
```cs
if(originalAssetPath.StartsWith("/"))
   ...
   
if(!originalAssetPath.StartsWith("avares://"))
   throw ...
```

kekekeks commented on src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/GroupTransformers/XamlIncludeGroupTransformer.cs:
```diff
@@ -0,0 +1,133 @@
+                node);
+        }
+        
+        AssetLoader.RegisterResUriParsers();
```
I believe this should happen somewhere in MSBuild task initialization rather than in a transformer. I'm also not quite sure if that's a good idea to pollute MSBuild process with our global state

maxkatz6 replied:
I will move. Also I don't use avalonia global state here, don't access AssetLoader APIs. Instead only URI scheme configuration is registered in .NET URI parser. Otherwise it will lower case all the links.

kekekeks reviewed (approved):
Mostly lgtm

maxkatz6 committed:
Merge branch 'master' into xaml-group-transfomers

maxkatz6 committed:
Introduce a new NewServiceProviderNode

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0026685-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

maxkatz6 committed:
Other changes after review

maxkatz6 committed:
Merge remote-tracking branch 'origin/master' into xaml-group-transfomers

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0026696-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

wieslawsoltes commented:
@maxkatz6 Testing this and got following error:
```
App.axaml(7,6): Error AVLN:0004 Avalonia: Unable to resolve build method "Build:/Themes/DockFluentTheme.axaml" resource on the "Dock.Avalonia" assembly. Line 7, position 6.
```

Xaml:
```xaml
<StyleInclude Source="avares://Dock.Avalonia/Themes/DockFluentTheme.axaml" />
```

kekekeks commented:
Are you using reference assemblies? I believe those can't work with compile-time includes since required type doesn't exist in the reference assembly and we aren't patching it yet.

wieslawsoltes commented:
> Are you using reference assemblies? I believe those can't work with compile-time includes since required type doesn't exist in the reference assembly and we aren't patching it yet.

Don't think so, the `Dock.Avalonia` project contains `DockFluentTheme.axaml` and is directly referenced by sample app
https://github.com/wieslawsoltes/Dock/blob/Update/src/Dock.Avalonia/Themes/DockFluentTheme.axaml

kekekeks commented:
@maxkatz6 might be some race condition related to this: https://github.com/AvaloniaUI/Avalonia/blob/ed2a4a3d00ddc02bd86d06e15a9f1c7b757f9b45/src/Avalonia.Build.Tasks/CompileAvaloniaXamlTask.cs#L39-L40

wieslawsoltes commented:
Btw is this now required for theme files based on `Styles` ?
https://github.com/AvaloniaUI/Avalonia/blob/fc0aa01c57cc9efe53d33e0053d3acf3c1984aef/src/Avalonia.Themes.Fluent/FluentTheme.xaml.cs#L36

wieslawsoltes commented:
Another issue (change?) I found:

When I had in xaml file in same assembly:
```xaml
<ResourceInclude Source="avares://Dock.Avalonia/Controls/DocumentTabStripItem.axaml" />
```
and changed to 
```xaml
<ResourceInclude Source="/Controls/DocumentTabStripItem.axaml" />
```

I got error
```bash
Error AVLN:0004 Avalonia: Unable to resolve build method "Build:/Themes/Controls/DocumentTabStripItem.axaml" resource on the "Dock.Avalonia" assembly. Line 8, position 10.
```

I needed to add to `csproj` the following to make it work:
```xml
<AvaloniaResource Include="**/*.axaml" />
```

maxkatz6 commented:
> @maxkatz6 might be some race condition related to this:

Seems like so. When I build same project second time - it works just fine.

> Btw is this now required for theme files based on Styles ?

No, if theme class doesn't have a ctor, this line will be injected. But I needed a ctor to initialize these resources, so had to add this line manually. 

> When I had in xaml file in same assembly: and changed to

Found the problem. Fixing it.

> I needed to add to csproj the following to make it work:

I don't think it's needed, otherwise you might end up with duplicate avalonia resources. Next error is a result.

kekekeks commented:
>Seems like so. When I build same project second time - it works just fine.

Try changing `AfterTargets="AfterCompile"` to `AfterTargets="CoreCompile"`.

kekekeks commented:
Changing the output path just before CoreCompile and reverting it afterwards might work too.

wieslawsoltes commented:
> Another issue (change?) I found:
> 
> When I had in xaml file in same assembly:
> 
> ```
> <ResourceInclude Source="avares://Dock.Avalonia/Controls/DocumentTabStripItem.axaml" />
> ```
> 
> and changed to
> 
> ```
> <ResourceInclude Source="/Controls/DocumentTabStripItem.axaml" />
> ```
> 
> I got error
> 
> ```shell
> Error AVLN:0004 Avalonia: Unable to resolve build method "Build:/Themes/Controls/DocumentTabStripItem.axaml" resource on the "Dock.Avalonia" assembly. Line 8, position 10.
> ```
> 
> I needed to add to `csproj` the following to make it work:
> 
> ```
> <AvaloniaResource Include="**/*.axaml" />
> ```
> 
> Edit: but that (adding AvaloniaResource to csproj) caused another error:
> 
> ```
> Error AVLN:0002 Avalonia: Duplicate x:Class directive, Dock.Avalonia.Themes.DockFluentTheme is already used in /Themes/DockFluentTheme.axaml
> ```

Tested https://github.com/AvaloniaUI/Avalonia/pull/9571 and it's fixed.

wieslawsoltes commented:
> @maxkatz6 Testing this and got following error:
> 
> ```
> App.axaml(7,6): Error AVLN:0004 Avalonia: Unable to resolve build method "Build:/Themes/DockFluentTheme.axaml" resource on the "Dock.Avalonia" assembly. Line 7, position 6.
> ```
> 
> Xaml:
> 
> ```
> <StyleInclude Source="avares://Dock.Avalonia/Themes/DockFluentTheme.axaml" />
> ```

@maxkatz6 
Tested https://github.com/AvaloniaUI/Avalonia/pull/9571 and still get this error.

wieslawsoltes commented:
> > @maxkatz6 Testing this and got following error:
> > ```
> > App.axaml(7,6): Error AVLN:0004 Avalonia: Unable to resolve build method "Build:/Themes/DockFluentTheme.axaml" resource on the "Dock.Avalonia" assembly. Line 7, position 6.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Xaml:
> > ```
> > <StyleInclude Source="avares://Dock.Avalonia/Themes/DockFluentTheme.axaml" />
> > ```
> 
> @maxkatz6 Tested #9571 and I still get this error.

Interesting, when I switch to 
```
<DockFluentTheme />
```
instead of 
```
<StyleInclude Source="avares://Dock.Avalonia/Themes/DockFluentTheme.axaml" />
```
it works, so still some issue with `StyleInclude`.

wieslawsoltes commented:
> > > @maxkatz6 Testing this and got following error:
> > > ```
> > > App.axaml(7,6): Error AVLN:0004 Avalonia: Unable to resolve build method "Build:/Themes/DockFluentTheme.axaml" resource on the "Dock.Avalonia" assembly. Line 7, position 6.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Xaml:
> > > ```
> > > <StyleInclude Source="avares://Dock.Avalonia/Themes/DockFluentTheme.axaml" />
> > > ```
> > 
> > 
> > @maxkatz6 Tested #9571 and I still get this error.
> 
> Interesting, when I switch to
> 
> ```
> <DockFluentTheme />
> ```
> 
> instead of
> 
> ```
> <StyleInclude Source="avares://Dock.Avalonia/Themes/DockFluentTheme.axaml" />
> ```
> 
> it works, so still some issue with `StyleInclude`.

Fore reference this is generated app code:
```c#
// Decompiled with JetBrains decompiler
// Type: AvaloniaDemo.Xaml.App
// Assembly: DemoXaml, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null
// MVID: 710CC489-2D33-44F9-A635-313141D04CE2
// Assembly location: C:\Users\Administrator\Documents\GitHub\Dock\samples\DemoXaml\bin\Debug\net7.0\DemoXaml.dll
// Compiler-generated code is shown

using Avalonia;
using Avalonia.Controls;
using Avalonia.Controls.ApplicationLifetimes;
using Avalonia.Layout;
using Avalonia.Markup.Xaml.Styling;
using Avalonia.Markup.Xaml.XamlIl.Runtime;
using Avalonia.Styling;
using Avalonia.Themes.Fluent;
using CompiledAvaloniaXaml;
using System;
using System.Runtime.InteropServices;

namespace AvaloniaDemo.Xaml
{
  public class App : Application
  {
    private static Action<object> \u0021XamlIlPopulateOverride;

    public override void Initialize()
    {
      App.\u0021XamlIlPopulateTrampoline(this);
    }

    public override void OnFrameworkInitializationCompleted()
    {
      if (this.ApplicationLifetime is IClassicDesktopStyleApplicationLifetime applicationLifetime1)
        applicationLifetime1.MainWindow = (Window) new MainWindow();
      if (this.ApplicationLifetime is ISingleViewApplicationLifetime applicationLifetime2)
        applicationLifetime2.MainView = (Control) new MainView();
      base.OnFrameworkInitializationCompleted();
    }

    public App()
    {
      base.\u002Ector();
    }

    static void \u0021XamlIlPopulate([In] IServiceProvider obj0, [In] App obj1)
    {
      XamlIlContext.Context<App> context = new XamlIlContext.Context<App>(obj0, new object[1]
      {
        (object) \u0021AvaloniaResources.NamespaceInfo\u003A\u002FApp\u002Eaxaml.Singleton
      }, "avares://DemoXaml/App.axaml");
      context.RootObject = obj1;
      context.IntermediateRoot = (object) obj1;
      App app1;
      App app2 = app1 = obj1;
      context.PushParent((object) app2);
      App app3 = app2;
      app3.Name = "Dock Avalonia Demo";
      Styles styles1 = app3.Styles;
      FluentTheme fluentTheme = new FluentTheme();
      fluentTheme.Mode = FluentThemeMode.Light;
      styles1.Add((IStyle) fluentTheme);
      Styles styles2 = app3.Styles;
      StyleInclude styleInclude = new StyleInclude((IServiceProvider) context);
      styleInclude.Source = new Uri("avares://Dock.Avalonia/Themes/DockFluentTheme.axaml", UriKind.RelativeOrAbsolute);
      styles2.Add((IStyle) styleInclude);
      Styles styles3 = app3.Styles;
      Style style = new Style();
      style.Selector = ((Selector) null).OfType(typeof (TextBlock));
      Setter setter1 = new Setter();
      setter1.Property = (AvaloniaProperty) Layoutable.HorizontalAlignmentProperty;
      setter1.Value = (object) HorizontalAlignment.Center;
      style.Add((ISetter) setter1);
      Setter setter2 = new Setter();
      setter2.Property = (AvaloniaProperty) Layoutable.VerticalAlignmentProperty;
      setter2.Value = (object) VerticalAlignment.Center;
      style.Add((ISetter) setter2);
      styles3.Add((IStyle) style);
      context.PopParent();
      if (app1 is StyledElement styled)
        NameScope.SetNameScope(styled, context.AvaloniaNameScope);
      context.AvaloniaNameScope.Complete();
    }

    private static void \u0021XamlIlPopulateTrampoline([In] App obj0)
    {
      if (App.\u0021XamlIlPopulateOverride != null)
        App.\u0021XamlIlPopulateOverride((object) obj0);
      else
        App.\u0021XamlIlPopulate(XamlIlRuntimeHelpers.CreateRootServiceProviderV2(), obj0);
    }
  }
}
```

maxkatz6 commented:
@wieslawsoltes 

> Fore reference this is generated app code:

Problem was with ref assembly, in which "!XamlIlPopulate" was not generated and was not copied from the non-ref assembly.
This method was used only for validation, that this class is a XAML class.
Removed this validation, and replaced with type check (i.e., type should implement IResourceDictionary or IStyle depending on Loaded property type).

wieslawsoltes commented:
> @wieslawsoltes
> 
> > Fore reference this is generated app code:
> 
> Problem was with ref assembly, in which "!XamlIlPopulate" was not generated and was not copied from the non-ref assembly. This method was used only for validation, that this class is a XAML class. Removed this validation, and replaced with type check (i.e., type should implement IResourceDictionary or IStyle depending on Loaded property type).

https://github.com/AvaloniaUI/Avalonia/pull/9571 fixed all issues I had, amazing work @maxkatz6
